Fix data race between AddModuleLevel and GetLevel

Asalle opened the pull request:
If not locked, this results in race:
1. SetBackend(), backend.go:26 - reassigning the global variable (write)
```
func SetBackend(backends ...Backend) LeveledBackend {
	var backend Backend
  // ...

  // DARA RACE START
	defaultBackend = AddModuleLevel(backend)
  // DATA RACE END

	return defaultBackend
}
```

2. IsEnabledFor(), logger.go:141 - accessing a default variable (read)
```
func (l *Logger) IsEnabledFor(level Level) bool {
	return defaultBackend.IsEnabledFor(level, l.Module) // DATA RACE HERE
}
```

zonque reviewed (changes requested):
Given that this is potential upstream material (as we can't afford the maintenance burden of a fork off this package), you need to explain where exactly the race is and also provide a commit log that justifies your change.

Asalle commented:
I'm in the process of testing it, will prepare a proper commit message as soon as it works

Commit:
backend, logger: lock before reassigning `defaultBackend`

If not locked, this results in race:
1. SetBackend(), backend.go:26 - reassigning the global variable (write)
```
func SetBackend(backends ...Backend) LeveledBackend {
	var backend Backend
  // ...

  // DARA RACE START
	defaultBackend = AddModuleLevel(backend)
  // DATA RACE END

	return defaultBackend
}
```

2. IsEnabledFor(), logger.go:141 - accessing a default variable (read)
```
func (l *Logger) IsEnabledFor(level Level) bool {
	return defaultBackend.IsEnabledFor(level, l.Module) // DATA RACE HERE
}
```

zonque commented:
> If not locked, this results in race:
> 
>     1. backend.go:27 - reassigning the global variable (write)
> 
>     2. logger.go:141 - accessing a default variable (read)

For the upstream PR, I suggest you quote some code here and name the variables by name. This will enable readers to comprehend things after the merge, when line numbers might have changed etc.

Asalle commented:
created a pr in upstream, closing this one
we'll use another fix
